### ISSUE TITLE
Possible for 200 workers to share 4 connections through pooling?

### ISSUE THREAD

**Issue by colinblaise** [frankhommers/Hangfire.PostgreSql#156]
I'd like to maintain 200 hangfire workers with the minimal # of db connections.

I see tons of errors on startup when using 200 worker count and using pooling and setting the db limit to 4 connections.

I saw that issue https://github.com/frankhommers/Hangfire.PostgreSql/issues/140 was closed and merged into version 1.7.0, which seems relevant to what I am seeing.

I am using the following postgres connection string:
`Host=127.0.0.1;Database=somedb;Username=someuser;Password=somepw;Pooling=true;Maximum Pool Size=4;Timeout=30`

I set Hangfire WorkerCount to 200.

On startup with no data, I receive 200 of the following:
```
[16:29:09 DBG Hangfire.Processing.BackgroundExecution] Execution loop Worker:92ea93eb has started in 961.7289 ms
  {}
[16:29:09 DBG Hangfire.Processing.BackgroundExecution] Execution loop Worker:d27eff49 has started in 983.4526 ms
  {}
[16:29:09 DBG Hangfire.Processing.BackgroundExecution] Execution loop Worker:1c805503 has started in 1015.2144 ms

...
```

After 30 seconds, I receive a large amount of the following:
```
[16:30:34 DBG Hangfire.Processing.BackgroundExecution] Execution loop Worker:7c8e10ff caught an exception and will be retried in 00:05:00
  {}
Npgsql.NpgsqlException (0x80004005): The connection pool has been exhausted, either raise MaxPoolSize (currently 4) or Timeout (currently 30 seconds)
   at Npgsql.ConnectorPool.AllocateLong(NpgsqlConnection conn, NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken)
   at Npgsql.NpgsqlConnection.<>c__DisplayClass32_0.<<Open>g__OpenLong|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Npgsql.NpgsqlConnection.Open()
   at Hangfire.PostgreSql.PostgreSqlStorage.CreateAndOpenConnection()
   at Hangfire.PostgreSql.PostgreSqlStorage.GetConnection()
   at Hangfire.Server.Worker.Execute(BackgroundProcessContext context)
   at Hangfire.Server.BackgroundProcessDispatcherBuilder.ExecuteProcess(Guid executionId, Object state)
   at Hangfire.Processing.BackgroundExecution.Run(Action`2 callback, Object state)
[16:30:34 DBG Hangfire.Processing.BackgroundExecution] Execution loop Worker:7c8e10ff will be retried in 00:05:00...
```

Am I crazy for thinking what I am trying to do should work?

**Comment by vytautask** [comment]
>Am I crazy for thinking what I am trying to do should work?

From my point of view - this works as designed and intended. That's a debug message innit?

**Comment by boukeversteegh** [comment]
It seems that #140 was actually not fixed by release 1.7.

I can see 1 connection for each configured worker. 25 workers = 25 connections. The pooling setting is ignored.

On one hand I see comments that say this is by design, on the other hand it was acknowledged as a bug and reportedly fixed, but the issue is still there.

Having the ability to control the number of connections separately from workers would be very useful, and I can imagine essential for anyone using Hangfire for parallel job processing. Hope someone is interested in fixing this.

For now, I'm setting my worker count to 1, as I currently don't need many workers and my connections are limited.

**Comment by azygis** [comment]
The connection management has been rewritten with release 1.9 and no connections are being held indefinitely by the library itself (like it has been previously, [as this comment says](https://github.com/frankhommers/Hangfire.PostgreSql/issues/140#issuecomment-607293455)). Whatever else is happening under the hood, in Npgsql, we do not really control.

It would be great if you try that again using the latest package and report back if the issue persists. Closing the issue in the meantime.
